compiler/rustc_errors/src/diagnostic_builder.rs

```diff
@@ -99,20 +99,22 @@ impl<'a> DerefMut for DiagnosticBuilder<'a> {
 
 impl<'a> DiagnosticBuilder<'a> {
     /// Emit the diagnostic.
-    pub fn emit(&mut self) {
+    pub fn emit(&mut self) -> ErrorReported {
```

Comment:
This needs to be `Option` since the diagnostic might not have been an error. I believe you may want to add a `type/`const` parameter to `DiagnosticBuilder` to tell apart error from non-error, and allow this method to always return `ErrorReported`, if the builder was created for an error in the first place.